crucible-wasm/src/Lang/Crucible/Wasm/Instantiate.hs

```diff
@@ -0,0 +1,531 @@
+  updStore off st =
+    do (Wasm.TableType (Wasm.Limit lmin _) Wasm.AnyFunc, addr) <- resolveTableIndex tableIndex im
+       functab <- Seq.lookup addr (storeTables st)
+       guard (toInteger off + toInteger (length funcIndexes) <= toInteger lmin)
+       funaddrs <- mapM (\i -> resolveFuncIndex i im) funcIndexes
+       functab' <- foldM (updTable st) functab (zip [ off .. ] funaddrs)
+       pure st{ storeTables = Seq.update addr functab' (storeTables st) }
+
+
+computeGlobalValue ::
+  ModuleInstance ->
+  Wasm.Global ->
+  InstM ConstantValue
+computeGlobalValue i (Wasm.Global{ globalType = t, initializer = expr }) =
+  do let i' = emptyInstance{ instGlobalAddrs = instGlobalAddrs i }
+     c <- evalConst i' expr
+     case t of
+        Wasm.Const vt -> checkConstantType c vt
+        Wasm.Mut vt -> checkConstantType c vt
+     return c
+
+checkConstantType :: ConstantValue -> Wasm.ValueType -> InstM ()
+checkConstantType _ _ = return () -- TODO!
+
+evalConst ::
+  ModuleInstance ->
+  Wasm.Expression ->
+  InstM ConstantValue
+evalConst _i [Wasm.I32Const x] = pure (I32Val x)
+evalConst _i [Wasm.I64Const x] = pure (I64Val x)
+evalConst _i [Wasm.F32Const x] = pure (F32Val x)
+evalConst _i [Wasm.F64Const x] = pure (F64Val x)
+
+evalConst i [Wasm.GetGlobal idx] =
+  do st <- lift get
+     case resolveGlobalIndex idx i of
+       Just (_,_,addr) | Just gv <- Seq.lookup addr (storeGlobals st)  ->
+         case gv of
+           ConstantGlobal v -> pure v
+           _ -> instErr ("global index not a constant " <> fromString (show idx))
+       _ -> instErr ("could not resolve global index " <> fromString (show idx))
+
+evalConst _ _ = instErr "expression not a constant!"
+
+bindImport :: Namespace -> ModuleInstance -> Wasm.Import -> InstM ExternalValue
+bindImport ns i Wasm.Import{ .. } =
+  case namespaceLookup (Wasm.Ident sourceModule) name ns of
+    Just (et,a)
+      | Just et' <- matchImportDesc i desc et -> pure (et',a)
+      | otherwise -> instErr ("Entity did not have expected type" <> sourceModule <> " " <> name)
+    Nothing -> instErr ("Could not lookup entity " <> sourceModule <> " " <> name)
+
+matchImportDesc :: ModuleInstance -> Wasm.ImportDesc -> ExternalType -> Maybe ExternalType
+matchImportDesc i (Wasm.ImportFunc tidx) (ExternalFunc t)
+  | Just t' <- resolveTypeIndex tidx i, t == t' = Just (ExternalFunc t)
+  | otherwise = Nothing
+
+matchImportDesc _ (Wasm.ImportGlobal gt) (ExternalGlobal gt')
+  | gt == gt' = Just (ExternalGlobal gt)
+
+matchImportDesc _ (Wasm.ImportTable (Wasm.TableType lim tt))
+                    (ExternalTable (Wasm.TableType lim' tt'))
+  | tt == tt' && matchLim lim lim' = Just (ExternalTable (Wasm.TableType lim tt))
+
+matchImportDesc _ (Wasm.ImportMemory mt) (ExternalMem (Wasm.Memory mt'))
+  | matchLim mt mt' = Just (ExternalMem (Wasm.Memory mt))
+
+matchImportDesc _ _ _ = Nothing
+
+-- TODO double check that we got this limit checking stuff in the right order
+--  and it all makes sense
```

Comment:
Looks reasonable to me.  Is the TODO still needed?  If it's still needed, can you provide more here about the context in which it needs to be validated?

Reply:
There's some slightly complicated contravariance going on here, and I want to really check carefully against the spec that I got all the inequalities in the correct direction.

Reply:
Can you add this context to the comments in the code?